Send metrics from a de-scheduled check

vickenty opened the pull request:
### What does this PR do?

Postpone removing the check sampler for a de-scheduled check until after we have handled and flushed all metrics from this check.

### Motivation

Avoid gaps when long-running cluster checks are rebalanced.

### Additional Notes

### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes

Run the agent with 
```
DD_AUTOCONF_CONFIG_FILES_POLL=true
DD_AUTOCONF_CONFIG_FILES_POLL_INTERVAL=30
DD_LOG_LEVEL=debug
DD_LOG_PAYLOADS=true
```

Add a python check that sends metrics after it is cancelled, set `empty_default_hostname` instance parameter to true.  Example: https://gist.github.com/vickenty/17fe1912957212184202c9ce4eef565f (This example relies on config polling to let the check schedule and de-schedule itself).

Check that the check.metric appears in the logs, and has an empty `host` value.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

vickenty committed:
Move telemetry updates into handler methods

Make a single method responsible for everything, so we can refactor
the channel code without impacting functionality.

vickenty committed:
Send check metrics and sketches via the same channel

Make samples, sketches and commit messages be handled by the
aggregator in the same linear order they were sent by the sender.

This fixes a small race where sketches sent before Commit() could be
handled after the Commit() was applied to the sampler (see
sender_test.go change around 467), delaying the sketch delivery
slightly.

vickenty committed:
Keep the check sampler alive until it is flushed

When a check is deregistered, we used to immediately delete the check
sampler, dropping all metrics sent by the check since the last flush.

To reliably send metrics from a check we need to keep the sampler as
long as there is potential for it to receive metrics, and then again
until it is flushed.

This patch adds a new message type, deregisterSampler, sent via the
metrics channel, to indicate that the sampler will not receive any
more metrics. Due to FIFO nature of the channel, it will be handled
after all the metrics sent by the check. At this point we know no more
metrics will arrive and can drop the sampler once we flush the
accumulated metrics. This is achieved with a flag, which is raised
when the aggregator processes deregisterSampler message, and the flag
is checked during flush: if the flag is set we know that no more
metrics will arrive for this particular check sampler.

vickenty committed:
Fix race in the aggregator test

IsInputQueueEmpty() is not sufficient, as the message may have
received from the queue, but did not have the effect we want. Instead
of checking the channel, wait for the flag change directly.

vickenty committed:
Fix data race in the test

This test now runs in parallel with the main aggregator thread, so we
need to use locks when poking at the aggregator internals.

vickenty committed:
Merge branch 'main' into seq-drop

ogaca-dd commented on pkg/aggregator/aggregator.go:
```diff
@@ -877,3 +879,41 @@ func (agg *BufferedAggregator) updateChecksTelemetry() {
+}
+
+func (s *deregisterSampler) handle(agg *BufferedAggregator) {
+	agg.mu.Lock()
+	defer agg.mu.Unlock()
+	if cs, ok := agg.checkSamplers[s.id]; ok {
+		cs.deregistered = true
+	}
```
Can you create a function for these lines in `BufferedAggregator` in order to encapsulate this logic inside `*BufferedAggregator`?

ogaca-dd reviewed (changes requested):
The PR looks good but I think the whole code for checks is too complex and would require some cleanups (Outside of the scope of this PR).

vickenty committed:
Move code to improve incapsulation

vickenty committed:
Merge remote-tracking branch 'origin/main' into seq-drop

vickenty committed:
Make test utility method available for tests only

vickenty committed:
Update collector tests

Update tests for cancelled checks to verify that metrics sent during
or immediately before the cancel are sent correctly.

vickenty committed:
Merge remote-tracking branch 'origin/main' into seq-drop
